vowpalwabbit/core/include/vw/core/reductions/epsilon_decay.h

```diff
@@ -22,51 +22,62 @@ namespace epsilon_decay
 struct epsilon_decay_score : scored_config
 {
   epsilon_decay_score() = default;
-  epsilon_decay_score(double alpha, double tau, uint64_t model_idx)
-      : VW::scored_config(alpha, tau), _model_idx(model_idx)
-  {
-  }
+  epsilon_decay_score(double alpha, double tau) : VW::scored_config(alpha, tau) {}
   float decayed_epsilon(uint64_t update_count);
   float get_upper_bound() const { return this->current_ips(); }
   float get_lower_bound() const { return _lower_bound; }
-  uint64_t get_model_idx() const { return _model_idx; }
+  uint64_t get_score_idx() const { return _score_idx; }
   void update_bounds(float w, float r);
+  void reset_stats(double alpha = DEFAULT_ALPHA, double tau = DEFAULT_TAU);
 
   float _lower_bound = 0.f;
-  uint64_t _model_idx;
+  uint64_t _score_idx;
 };
 
 struct epsilon_decay_data
 {
   epsilon_decay_data(uint64_t num_configs, uint64_t min_scope, double epsilon_decay_alpha, double epsilon_decay_tau,
-      parameters& weights, VW::io::logger logger, bool log_champ_changes)
+      dense_parameters& weights, VW::io::logger logger, bool log_champ_changes, bool constant_epsilon)
       : _min_scope(min_scope)
       , _epsilon_decay_alpha(epsilon_decay_alpha)
       , _epsilon_decay_tau(epsilon_decay_tau)
       , _weights(weights)
       , _logger(std::move(logger))
       , _log_champ_changes(log_champ_changes)
+      , _constant_epsilon(constant_epsilon)
   {
-    uint64_t model_idx = 0;
+    uint64_t weight_idx = 0;
     for (uint64_t i = 0; i < num_configs; ++i)
     {
       std::vector<epsilon_decay_score> score_vec;
       for (uint64_t j = 0; j < i + 1; ++j)
       {
-        epsilon_decay_score s(epsilon_decay_alpha, epsilon_decay_tau, model_idx);
+        epsilon_decay_score s(epsilon_decay_alpha, epsilon_decay_tau);
         score_vec.push_back(s);
```

Comment:
```suggestion
        score_vec.emplace_back(epsilon_decay_alpha, epsilon_decay_tau);
```

Reply:
there's no reason to create, copy, then destroy this object in every loop iteration

Reply:
updated